# Introduction
This example looks at the Kaggle Credit Card Fraud Detection dataset to demonstrate how to train a classification model on data with highly imbalanced classes.

In [1]:
import numpy as np
import csv

# Get the real data from https://www.kaggle.com/mlg-ulb/creditcardfraud/
fname = "../data/imbalanced-classification/creditcard.csv"

all_features = []
all_targets = []
with open(fname) as f:
    for i, line in enumerate(f):
        if i == 0:
            print("HEADER:", line.strip())
            continue # skip header
        fields = line.strip().split(",")
        all_features.append([float(v.replace('"', "")) for v in fields[:-1]])
        all_targets.append([int(fields[-1].replace('"', ""))])
        if i == 1:
            print("EXAMPLE FEATURES:", all_features[-1])
features = np.array(all_features, dtype="float32")
targets = np.array(all_targets, dtype="uint8")
print("features.shape:", features.shape)
print("targets.shape:", targets.shape)

HEADER: "Time","V1","V2","V3","V4","V5","V6","V7","V8","V9","V10","V11","V12","V13","V14","V15","V16","V17","V18","V19","V20","V21","V22","V23","V24","V25","V26","V27","V28","Amount","Class"
EXAMPLE FEATURES: [0.0, -1.3598071336738, -0.0727811733098497, 2.53634673796914, 1.37815522427443, -0.338320769942518, 0.462387777762292, 0.239598554061257, 0.0986979012610507, 0.363786969611213, 0.0907941719789316, -0.551599533260813, -0.617800855762348, -0.991389847235408, -0.311169353699879, 1.46817697209427, -0.470400525259478, 0.207971241929242, 0.0257905801985591, 0.403992960255733, 0.251412098239705, -0.018306777944153, 0.277837575558899, -0.110473910188767, 0.0669280749146731, 0.128539358273528, -0.189114843888824, 0.133558376740387, -0.0210530534538215, 149.62]
features.shape: (284807, 30)
targets.shape: (284807, 1)


# Prepare a validation set

In [2]:
num_val_samples = int(len(features) * 0.2)
train_features = features[:-num_val_samples]
train_targets = targets[:-num_val_samples]
val_features = features[-num_val_samples:]
val_targets = targets[-num_val_samples:]

print("Number of training samples:", len(train_features))
print("Number of validation samples:", len(val_features))

Number of training samples: 227846
Number of validation samples: 56961


# Analyze class imbalance in the targets

In [3]:
counts = np.bincount(train_targets[:, 0])
for i, count in enumerate(counts):
    print("# of count for class {}: {}".format(i, count))

# of count for class 0: 227429
# of count for class 1: 417


In [4]:
print(
    "Number of positive samples in training data: {} ({:.2f}% of total)".format(
        counts[1], 100 * float(counts[1]) / len(train_targets)
    )
)

weight_for_0 = 1.0 / counts[0]
weight_for_1 = 1.0 / counts[1]

print("weight for class 0: {}".format(weight_for_0))
print("weight for class 1: {}".format(weight_for_1))

Number of positive samples in training data: 417 (0.18% of total)
weight for class 0: 4.396976638863118e-06
weight for class 1: 0.002398081534772182


# Normalize the data using ***training*** set statistics

In [5]:
mean = np.mean(train_features, axis=0)
train_features -= mean
val_features -= mean
std = np.std(train_features, axis=0)
train_features /= std
val_features /= std

# Build a binary classification model

In [6]:
from tensorflow import keras

l = keras.layers

model = keras.Sequential([
    l.Dense(256, activation='relu', input_shape=(train_features.shape[-1],)),
    l.Dense(256, activation='relu'),
    l.Dropout(0.3),
    l.Dense(256, activation='relu'),
    l.Dropout(0.3),
    l.Dense(1, activation='sigmoid'),
])

model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 256)               7936      
_________________________________________________________________
dense_1 (Dense)              (None, 256)               65792     
_________________________________________________________________
dropout (Dropout)            (None, 256)               0         
_________________________________________________________________
dense_2 (Dense)              (None, 256)               65792     
_________________________________________________________________
dropout_1 (Dropout)          (None, 256)               0         
_________________________________________________________________
dense_3 (Dense)              (None, 1)                 257       
Total params: 139,777
Trainable params: 139,777
Non-trainable params: 0
__________________________________________________

# Train the model with `class_weight` argument

In [8]:
m = keras.metrics

metrics = [
    m.FalseNegatives(name='fn'),
    m.FalsePositives(name='fp'),
    m.TrueNegatives(name='tn'),
    m.TruePositives(name='tp'),
    m.Precision(name='precision'),
    m.Recall(name='recall'),
]

model.compile(optimizer=keras.optimizers.Adam(1e-2),
              loss='binary_crossentropy',
              metrics=metrics)

callbacks = [keras.callbacks.ModelCheckpoint("../checkpoints/imbalanced-classification/fraud_model_at_epoch_{epoch}.h5")]
class_weight = {0: weight_for_0, 1: weight_for_1}

model.fit(
    train_features,
    train_targets,
    batch_size=2048,
    epochs=30,
    verbose=2,
    callbacks=callbacks,
    validation_data=(val_features, val_targets),
    class_weight=class_weight,
)

Epoch 1/30
112/112 - 1s - loss: 2.2907e-07 - fn: 2.0000 - fp: 2718.0000 - tn: 224711.0000 - tp: 415.0000 - precision: 0.1325 - recall: 0.9952 - val_loss: 0.0133 - val_fn: 11.0000 - val_fp: 306.0000 - val_tn: 56580.0000 - val_tp: 64.0000 - val_precision: 0.1730 - val_recall: 0.8533
Epoch 2/30
112/112 - 0s - loss: 2.5064e-07 - fn: 5.0000 - fp: 2286.0000 - tn: 225143.0000 - tp: 412.0000 - precision: 0.1527 - recall: 0.9880 - val_loss: 0.0137 - val_fn: 13.0000 - val_fp: 195.0000 - val_tn: 56691.0000 - val_tp: 62.0000 - val_precision: 0.2412 - val_recall: 0.8267
Epoch 3/30
112/112 - 0s - loss: 1.5674e-07 - fn: 1.0000 - fp: 1657.0000 - tn: 225772.0000 - tp: 416.0000 - precision: 0.2007 - recall: 0.9976 - val_loss: 0.0138 - val_fn: 13.0000 - val_fp: 278.0000 - val_tn: 56608.0000 - val_tp: 62.0000 - val_precision: 0.1824 - val_recall: 0.8267
Epoch 4/30
112/112 - 0s - loss: 2.8296e-07 - fn: 1.0000 - fp: 1956.0000 - tn: 225473.0000 - tp: 416.0000 - precision: 0.1754 - recall: 0.9976 - val_loss: 

# Conclusions

At the end of training, out of 56,961 validation transactions, we are:

Correctly identifying 66 of them as fraudulent
Missing 9 fraudulent transactions
At the cost of incorrectly flagging 284 legitimate transactions
In the real world, one would put an even higher weight on class 1, so as to reflect that False Negatives are more costly than False Positives.

Next time your credit card gets declined in an online purchase -- this is why.